Src/Sarif.PatternMatcher.Function/SpamAnalyzer.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+using Microsoft.CodeAnalysis.Sarif.Driver;
+using Microsoft.CodeAnalysis.Sarif.Writers;
+
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Function
+{
+    internal static class SpamAnalyzer
+    {
+        internal static readonly IFileSystem FileSystem;
+
+        static SpamAnalyzer()
+        {
+            FileSystem = Sarif.FileSystem.Instance;
+        }
+
+        public static SarifLog Analyze(string filePath, string text, string rulePath)
+        {
+            IEnumerable<string> regexDefinitions = FileSystem.DirectoryGetFiles(Path.Combine(rulePath, @"..\Rules"), "*.json");
+
+            // Load all rules from JSON. This also automatically loads any validations file that
+            // lives alongside the JSON. For a JSON file named PlaintextSecrets.json, the
+            // corresponding validations assembly is named PlaintextSecrets.dll (i.e., only the
+            // extension name changes from .json to .dll).
+            ISet<Skimmer<AnalyzeContext>> skimmers =
+                AnalyzeCommand.CreateSkimmersFromDefinitionsFiles(FileSystem, regexDefinitions);
+
+            var sb = new StringBuilder();
+
+            using (var outputTextWriter = new StringWriter(sb))
+            using (var logger = new SarifLogger(
+                outputTextWriter,
+                LoggingOptions.PrettyPrint | LoggingOptions.Verbose))
+            {
+                // The analysis will disable skimmers that raise an exception. This
+                // hash set stores the disabled skimmers. When a skimmer is disabled,
+                // that catastrophic event is logged as a SARIF notification.
+                var disabledSkimmers = new HashSet<string>();
+
+                var context = new AnalyzeContext
+                {
+                    TargetUri = new Uri(filePath, UriKind.RelativeOrAbsolute),
+                    FileContents = text,
+                    Logger = logger,
+                };
+
+                using (context)
+                {
+                    IEnumerable<Skimmer<AnalyzeContext>> applicableSkimmers = AnalyzeCommand.DetermineApplicabilityForTargetHelper(context, skimmers, disabledSkimmers);
+                    AnalyzeCommand.AnalyzeTargetHelper(context, applicableSkimmers, disabledSkimmers);
+                }
+            }
+
+            SarifLog sarifLog = JsonConvert.DeserializeObject<SarifLog>(sb.ToString());
```

Comment:
>sarifLog [](start = 21, length = 8)

Right here you have the log file as JSON. You then serialize it to an actual SarifLog. Then in your function, you turn the sarif log back into JSON. Can we eliminate some of this round-tripping? e.g., can you just return sb.ToString() from this method and eliminate the JSON conversion upstream?

Reply:
I asked the same thing. the issue with the sb, is that it will contain the sarif as json with quotes everywhere....so, for now, i  said it was ok. but somehow we should try to not do that. (we are thinking the same thing)

---
In reply to: [553657933](https://github.com/microsoft/sarif-pattern-matcher/pull/85#discussion_r553657933) [](ancestors = 553657933)